backend/src/test/kotlin/app/ehrenamtskarte/backend/auth/webservice/ApiTokenServiceTest.kt

```diff
@@ -0,0 +1,172 @@
+import org.junit.Assert.assertFalse
+import org.junit.Assert.assertThrows
+import org.junit.Assert.assertTrue
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import java.time.LocalDate
+
+internal class ApiTokenServiceTest : IntegrationTest() {
+
+    private val mockDfe = mockk<DataFetchingEnvironment>()
+    private val mockContext = mockk<GraphQLContext>()
+    private val mockJwtPayload = mockk<JwtPayload>()
+
+    @BeforeEach
+    fun setUp() {
+        every { mockDfe.getContext<GraphQLContext>() } returns mockContext
+        every { mockContext.enforceSignedIn() } returns mockJwtPayload
+        transaction {
+            ApiTokens.deleteAll()
+        }
+    }
+
+    @Test
+    fun createApiToken_createsTokenSuccessfully() {
+        every { mockJwtPayload.adminId } returns TestAdministrators.KOBLENZ_PROJECT_ADMIN.id
+
+        transaction {
+            val numberOfTokensBefore = ApiTokens.selectAll().count()
+            ApiTokenService().createApiToken(3, mockDfe)
+            val numberOfTokensAfter = ApiTokens.selectAll().count()
+            assertEquals(numberOfTokensBefore + 1, numberOfTokensAfter)
+        }
+    }
+
+    @Test
+    fun createApiToken_throwsForbiddenExceptionWhenNotAuthorized() {
+        every { mockJwtPayload.adminId } returns TestAdministrators.NUERNBERG_PROJECT_STORE_MANAGER.id
+
+        transaction {
+            assertThrows(ForbiddenException::class.java) {
+                ApiTokenService().createApiToken(1, mockDfe)
+            }
+            assertFalse(
+                Authorizer.mayAddApiTokensInProject(
+                    AdministratorEntity.find { Administrators.id eq mockJwtPayload.adminId }.single()
+                )
+            )
+        }
+    }
+
+    @Test
+    fun createApiToken_throwsExceptionWhenAdminNotFound() {
+        every { mockJwtPayload.adminId } returns 1234
+
+        transaction {
+            assertThrows(ForbiddenException::class.java) {
+                ApiTokenService().createApiToken(1, mockDfe)
+            }
+        }
+    }
+
+    @Test
+    fun deleteApiToken_deletesTokenSuccessfully() {
+        every { mockJwtPayload.adminId } returns TestAdministrators.KOBLENZ_PROJECT_ADMIN.id
+
+        transaction {
+            ApiTokens.insert {
+                it[ApiTokens.id] = 1
+                it[projectId] = 1
+                it[creatorId] = TestAdministrators.KOBLENZ_PROJECT_ADMIN.id
+                it[tokenHash] = ByteArray(1)
+                it[expirationDate] = LocalDate.now()
+            }
+
+            val tokenExists = ApiTokens.select { ApiTokens.id eq 1 }.count() > 0
+            assertTrue(tokenExists)
+
+            ApiTokenService().deleteApiToken(1, mockDfe)
+
+            val tokenNoLongerExists = ApiTokens.select { ApiTokens.id eq 1 }.singleOrNull() != null
+            assertTrue(tokenNoLongerExists)
+        }
+    }
+
+    @Test
+    fun deleteApiToken_deletesFromOtherAdminTokenSuccessfully() {
```

Comment:
it should be possible only inside one project. between Koblenz and Nuernberg is should fail

Reply:
Yes correct, that is checked by the next test: deleteApiToken_deletesNoTokenFormOtherProject

Reply:
I mean, this test is named - “deletes token successfully”, but in this test the Koblenz admin tried to delete a token created by the Nürnberg admin, which should not be successful.
The test is currently green because of the same typo in the check condition as it was above: 
`val tokenNoLongerExists = ApiTokens.select { ApiTokens.id eq 1 }.singleOrNull() == null`
I think if we want to check that a token can be deleted by other admin, both admins (creating and deleting) should be from the same project.

Reply:
Hi Tory, I understand your confusion here.

The expected behavior is, that:
1) An admin can delete all tokens of his project (even tokens of other admins)
2) An admin cannot delete tokens of other projects

This test deletes the token successfully as expected, as the token is created for the koblenz project (by the nuernberg admin). This would not be possible of course in the real software, but as we do not have another koblenz admin i just used the nuernberg admin id here. But to avoid confusion i will create a second koblenz project admin

Reply:
> the token is created for the koblenz project (by the nuernberg admin)

in the latest version of the test, the token was created for the nuernberg project by the nurnberg admin.
and it wasn't removed at the end.
because `ApiTokens.select { ApiTokens.id eq 1 }.singleOrNull() != null` means, the token is not null, so it's still there

Reply:
ok thanks for explaining and you patience. I now fixed it.